husky_extensions/itemgrid.js

```diff
@@ -0,0 +1,311 @@
+                of: 'public.of',
+                visible: 'public.visible',
+                elementsSelected: 'public.elements-selected'
+            }
+        },
+
+        constants = {
+            noContentClass: 'no-content',
+            addIcon: 'fa-plus-circle',
+            deleteIcon: 'fa-trash-o'
+        },
+
+        /**
+         * returns the normalized event names
+         * @param eventName {string} The name of the concrete event without prefix
+         * @returns {string} Returns the prefixed event name
+         */
+        createEventName = function(eventName) {
+            return [
+                this.options.eventNamespace,
+                '.',
+                (this.options.instanceName ? this.options.instanceName + '.' : ''),
+                eventName
+            ].join('');
+        },
+
+        templates = {
+            skeleton: function() {
+                return [
+                    '<div class="listgrid" id="', this.ids.container, '">',
+                    '    <div class="header">',
+                    '        <span class="', constants.addIcon ,' icon left action add-button" id="', this.ids.addButton, '"></span>',
+                    '        <span class="', constants.deleteIcon ,' icon left action delete-button" id="', this.ids.deleteButton, '"></span>',
+                    '    </div>',
+                    '    <div class="content" id="', this.ids.content, '"></div>',
+                    '</div>'
+                ].join('');
+            }
+        },
+
+        bindCustomEvents = function() {
+            this.sandbox.on(this.DATA_CHANGED(), this.changeData.bind(this));
+
+            this.sandbox.on(this.DELETE_BUTTON_CLICKED(), function() {
+                this.sandbox.emit('husky.datagrid.' + this.options.instanceName + '.items.get-selected', function(ids) {
+                    deleteItems.call(this, ids);
+                }.bind(this));
+            }.bind(this));
+
+            this.sandbox.on(this.ITEM_ADD(), function(item) {
+                addItem.call(this, item);
+            }.bind(this));
+
+            this.sandbox.on(this.ITEM_SAVE(), function(item) {
+                editItem.call(this, item);
+            }.bind(this));
+        },
+
+        bindDomEvents = function() {
+            // click on the add button
+            this.sandbox.dom.on(this.$addButton, 'click', function() {
+                this.sandbox.emit(this.ADD_BUTTON_CLICKED());
+            }.bind(this));
+
+            // click on the config button
+            this.sandbox.dom.on(this.$deleteButton, 'click', function() {
+                this.sandbox.emit(this.DELETE_BUTTON_CLICKED());
+            }.bind(this));
+        },
+
+        addItem = function(newItem) {
+            var data = [];
+
+            if (App.util.typeOf(this.getData()) === 'array' && App.util.typeOf(this.getData()[0]) === 'object') {
+                data = this.getData();
+            }
```

Comment:
I don't really understand the logic here at all... I would write it something like this:

```javascript
var data = this.getData();

if (this.sandbox.util.typeOf(data) === ...) {
    data = [];
}
```

This way the `getData` method would only be called once, which is good, because I would assume that this method could also have a serious performance impact, depending on the type of list it is used with.